### PR TITLE
Update CHANGELOG for 3.7.0

### PR DESCRIPTION
**Pull request by jbech-linaro**
@OP-TEE/maintainers please start testing your devices/platforms and add `Tested-by: ...` in the comments. Release candidate tag will be sent out shortly.

I've also written a python script that makes it quick and easy to create/update the `CHANGELOG.md`. The changelog updates in this commit has been generated by that script. Another change is that it also lists the other gits that makes up the core of the OP-TEE project.

// Joakim

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,25 @@
+# OP-TEE - version 3.7.0 (2019-10-11)
+
+- Link to the GitHub [release page][github_release_3_7_0] (optee_os)
```
Comment (jforissier) [inline CHANGELOG.md]: Why only link to the release page of optee_os and not optee_client etc.?

File: scripts/update_changelog.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+                                     'the CHANGELOG.md file.\n'
+                                     'Usage example:\n'
+                                     '  ./update_changelog.py '
+                                     ' --changelog-file CHANGELOG.md'
+                                     ' --release-version 3.7.0'
+                                     ' --previous-release-version 3.6.0'
+                                     ' --release-date 2019-10-11'
+                                     ' --previous-release-date 2019-07-05')
+
+    parser.add_argument('--changelog-file', action='store', required=True,
```
[the remark this answers is not in the document]
Comment (jbech-linaro) [reply scripts/update_changelog.py]: Agree.

File: scripts/update_changelog.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+                        help='Changelog file to be updated.')
+
+    parser.add_argument('--release-date', action='store', required=True,
+                        help='The release date (yyyy-mm-dd).')
+
+    parser.add_argument('--previous-release-date', action='store',
```
[the remark this answers is not in the document]
Comment (jbech-linaro) [reply scripts/update_changelog.py]: Good idea, I'll update the script.

File: CHANGELOG.md (hunk 1 above)
Comment (jbech-linaro) [reply CHANGELOG.md]: I can do it, I'll update the script.

**Comment by jbech-linaro** [comment]
> Reviewing "scripts: add update_changelog.py".
> 
> I find it a bit weird to mention the other Gits here. 

I don't think so, OP-TEE is more than just optee_os, so when making releases it makes sense to include the other core gits also.

> In fact, `build.git` or `manifest.git` would probably be a better place for the change log I would think.

It's a legacy thing with having it in optee_os. IMHO, the changelog etc should be at `optee_docs`. But since we're short on time I suggest that we keep using current changelog in optee_os.

**Comment by jforissier** [comment]
> It's a legacy thing with having it in optee_os. IMHO, the changelog etc should be at optee_docs. But since we're short on time I suggest that we keep using current changelog in optee_os.

Yeah `optee_os` would be a good place, too. But I agree we can keep it in `optee_os` for now, it doesn't matter much.

File: scripts/update_changelog.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jbech-linaro) [reply scripts/update_changelog.py]: Done

File: scripts/update_changelog.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (jbech-linaro) [reply scripts/update_changelog.py]: Done.

File: CHANGELOG.md (hunk 1 above)
Comment (jbech-linaro) [reply CHANGELOG.md]: Done.

**Comment by jbech-linaro** [comment]
Comment to myself, need to update the commit message in the script patch, since that still mentions the previous date.

**Comment by jforissier** [comment]
LGTM, for "scripts: add update_changelog.py":
`Acked-by: Jerome Forissier <jerome@forissier.org>`

**Comment by bryanodonoghue** [comment]
`Tested-by: Bryan O'Donoghue <bryan.odonoghue@linaro.org>` WaRP7

**Comment by mimok** [comment]
`Tested-by: Michael Grand <michael.grand.mg@gmail.com>`  ZynqMP (zcu102, ultra96v1)

**Comment by Emantor** [comment]
`Tested-by: Rouven Czerwinski <r.czerwinski@pengutronix.de> (imx-imx8mqevk)`

**Comment by Emantor** [comment]
FWIW, gcc-9 emits:
```
core/drivers/crypto/caam/caam_jr.c: In function 'do_jr_dequeue':
core/drivers/crypto/caam/caam_jr.c:262:22: warning: taking address of packed member of 'struct outring_entry' may result in an unaligned pointer value [-Waddress-of-packed-member]
  262 |    if (caam_desc_pop(&jr_out->desc) == caller->pdesc) {
      |                      ^~~~~~~~~~~~~
core/drivers/crypto/caam/caam_jr.c:265:26: warning: taking address of packed member of 'struct outring_entry' may result in an unaligned pointer value [-Waddress-of-packed-member]
  265 |      caam_read_jobstatus(&jr_out->status);
      |                          ^~~~~~~~~~~~~~~
```

**Comment by jbech-linaro** [comment]
> FWIW, gcc-9 emits:

@Emantor , thanks for reporting. For this particular release, I'd say that we just ignore that for now. We're trying to follow the [gcc releases from Arm](https://developer.arm.com/tools-and-software/open-source-software/developer-tools/gnu-toolchain/gnu-a/downloads) and so far there are no gcc9 listed for download. Having said that, if you have time and want to fix the gcc9 issues already now, then we would of course not complain :wink:.

**Comment by b49020** [comment]
`Tested-by: Sumit Garg <sumit.garg@linaro.org> (Developerbox)`

**Comment by clementfaure** [comment]
We will do the testing this week.
@jbech-linaro I have just a question : is the CAAM driver included in this 3.7 release? I see that @Emantor compiled the CAAM driver code in a previous comment.

**Comment by jbech-linaro** [comment]
> @clementfaure is the CAAM driver included in this 3.7 release?

#3228 #3259 and #2962 are the big ones I guess for CAAM and they have all been merged which means that they will be part of 3.7.0. I.e., all this [here](https://github.com/OP-TEE/optee_os/pulls?utf8=%E2%9C%93&q=is%3Apr+is%3Aclosed+CAAM+) will be included. Have I missed anything that should (or shouldn't) be included?

**Comment by clementfaure** [comment]
> > @clementfaure is the CAAM driver included in this 3.7 release?
> 
> #3228 #3259 and #2962 are the big ones I guess for CAAM and they have all been merged which means that they will be part of 3.7.0. I.e., all this [here](https://github.com/OP-TEE/optee_os/pulls?utf8=%E2%9C%93&q=is%3Apr+is%3Aclosed+CAAM+) will be included. Have I missed anything that should (or shouldn't) be included?

I believe this PR does not include #3259

**Comment by jforissier** [comment]
@clementfaure don't worry, this PR doesn't really matter, its purpose is just to updeta the change log and gathher Tested-by's. All PRs that are merged into master as of now will be in the 3.7.0 release, because the release is a tag created on master.

**Comment by clementfaure** [comment]
> @clementfaure don't worry, this PR doesn't really matter, its purpose is just to update the change log and gather Tested-by's. All PRs that are merged into master as of now will be in the 3.7.0 release, because the release is a tag created on master.

Oh, okay! 
Thank you for your answer.

**Comment by jforissier** [comment]
With https://github.com/OP-TEE/optee_os/pull/3315 and https://github.com/OP-TEE/optee_os/pull/3317:

`Tested-by: Jerome Forissier <jerome@forissier.org> (HiKey960, GP)`

**Comment by jbech-linaro** [comment]
@igoropaniuk , do you think you can give Rpi3 a try? I get stuck when booting up kernel. I've bisected all the way back to 3.6.0 but still the same issue. I gave a clean 3.6.0 a try and that also fails on my side (which it didn't when we did the 3.6.0 release).

**Comment by jbech-linaro** [comment]
> @igoropaniuk , do you think you can give Rpi3 a try?

Stepping back "build.git" 3.6.0 makes it work again with latest on everything else. Strangely there aren't many changes at all in build.git when it comes to Rpi3, but now I have something to investigate.

**Comment by jbech-linaro** [comment]
> but now I have something to investigate.

It's `GETTY_PORT ... ` thing changed in https://github.com/OP-TEE/build/commit/7fbd6ce22a596d1370e6ffba83a7ecfe42cda447.

**Comment by jforissier** [comment]
> It's `GETTY_PORT ... ` thing changed in https://github.com/OP-TEE/build/commit/7fbd6ce22a596d1370e6ffba83a7ecfe42cda447.

:cry: 

I can't spot the bug in the commit. Can you check which value of `BR2_TARGET_GENERIC_GETTY_PORT` in out-br/.config, with and without the patch?

**Comment by jbech-linaro** [comment]
> I can't spot the bug in the commit.

Issue found and fixed, see https://github.com/OP-TEE/build/pull/390

**Comment by nodeax** [comment]
```
Tested-by: Akshay Bhat <akshay.bhat@timesys.com> (Atmel SAM)

# xtest -l 15
+-----------------------------------------------------
44094 subtests of which 0 failed
96 test cases of which 0 failed
0 test cases were skipped
TEE test application done!
```

**Comment by jbech-linaro** [comment]
With OP-TEE/build#390,
`Tested-by: Joakim Bech <joakim.bech@linaro.org>` (Rpi3b)
```bash
# xtest -l 15
+-----------------------------------------------------
44094 subtests of which 0 failed
96 test cases of which 0 failed
0 test cases were skipped
TEE test application done!
```

**Comment by jenswi-linaro** [comment]
`Tested-by: Jens Wiklander <jens.wiklander@linaro.org> (Juno)`

**Comment by glneo** [comment]
```
Tested-by: Andrew F. Davis <afd@ti.com> (plat-k3, plat-ti)
```

**Comment by jbech-linaro** [comment]
`Tested-by: Joakim Bech <joakim.bech@linaro.org>` (FVP)
```
# xtest
+-----------------------------------------------------
24105 subtests of which 0 failed
96 test cases of which 0 failed
0 test cases were skipped
TEE test application done!
```
I tested `xtest -l 15` also, but after 3 hours it still hadn't finished and I had to stop it. I'll maybe give it a try again and let it run over night or so.

**Comment by igoropaniuk** [comment]
Regarding Poplar, there is build issue because of wrong path to TF-A in `l-loader`  in `Makefile` (https://github.com/96boards-poplar/l-loader/blob/master/Makefile#L13):

```
start.S:37:10: fatal error: poplar_layout.h: No such file or directory
 #include "poplar_layout.h"
          ^~~~~~~~~~~~~~~~~
compilation terminated.
Makefile:70: recipe for target 'start.o' failed
make[1]: *** [start.o] Error 1
make[1]: Leaving directory '/home/xdev/linaro/reps/poplar.repo/l-loader'
Makefile:121: recipe for target 'l-loader' failed
make: *** [l-loader] Error 2
```

I see two possible ways to resolve this (let me know what you prefer):
1. Address this directly in https://github.com/96boards-poplar/l-loader (and as good justification is TF-A transition to the new name/location)
2. Keep old directory name for poplar setup (revert it in build.git and manifest).

With this fix:

```
+-----------------------------------------------------
24109 subtests of which 0 failed
97 test cases of which 0 failed
0 test cases were skipped
TEE test application done!

```

**Comment by jforissier** [comment]
@igoropaniuk how about setting `ARM_TRUSTED_FIRMWARE` in the caller of l-loader?

**Comment by igoropaniuk** [comment]
@jforissier 

https://github.com/OP-TEE/build/pull/392
https://github.com/96boards-poplar/l-loader/pull/3

no changes needed for `poplar.xml` manifest, as it points to the latest master of `l-loader.

**Comment by jforissier** [comment]
@igoropaniuk FYI, I met the build issue fixed by this patch in upstream U-Boot: https://gitlab.denx.de/u-boot/u-boot/commit/e0d20dc1521e74b82dbd69be53a048847798a90a

I don't know how the maintainers at https://github.com/96boards-poplar/u-boot.git manage backports from upstream, hopefully this will get fix someday.

**Comment by igoropaniuk** [comment]
@igoropaniuk 
Seems that https://github.com/96boards-poplar/u-boot.git is abandoded.

btw, last year I played a lot with the mainline U-boot on Poplar and haven't met any issues (AFAIK I reported one regression which Shawn quickly fixed). So IMHO it's pretty mature.

So, what do you think about switching to one of the release versions of mainline U-boot (I can try and test v2019.10, that was released two days ago).

**Comment by jforissier** [comment]
> So, what do you think about switching to one of the release versions of mainline U-boot (I can try and test v2019.10, that was released two days ago).

Sounds like a plan! Thanks.

**Comment by igoropaniuk** [comment]
@jforissier 
https://github.com/OP-TEE/manifest/pull/148

**Comment by igoropaniuk** [comment]
`Tested-by: Igor Opaniuk <igor.opaniuk@gmail.com> (Poplar)`

**Comment by jbech-linaro** [comment]
With @igoropaniuk's patch merged, which was the last patch I was waiting for, I think we're ready to create the rc1 tag.

**Comment by jforissier** [comment]
@jbech-linaro agreed. There is still https://github.com/linaro-swg/linux/issues/69 but it's hardly an OP-TEE issue and it is certainly not a requirement for the release.

**Comment by sahilnxp** [comment]
With optee: commit (34a08bec)

`Tested-by: Sahil Malhotra <sahil.malhotra@nxp.com> (LS1046A-RDB)`

```
# xtest -l15
+-----------------------------------------------------
44094 subtests of which 0 failed
96 test cases of which 0 failed
0 test cases were skipped
TEE test application done!
root@localhost:~#
```

**Comment by jbech-linaro** [comment]
FYI 3.7.0-rc1 has been created across `optee_*`, `build` and `manifest`.

**Comment by jforissier** [comment]
3.7.0-rc1:

`Tested-by: Jerome Forissier <jerome@forissier.org> (HiKey, GP)`

**Comment by etienne-lms** [comment]
3.7.0-rc1 + https://github.com/OP-TEE/optee_os/pull/3330

`Tested-by: Etienne Carriere <etienne.carriere@linaro.org> (stm32mp1)`

**Comment by jforissier** [comment]
3.7.0-rc1-1-g8027bd32:

`Tested-by: Jerome Forissier <jerome@forissier.org> (QEMUv8)`

**Comment by vchong** [comment]
xtest 1022 fails in Hikey AOSP. :(

```
* regression_1022 Test dlopen()/dlsym()/dlclose() API
external/optee_test/host/xtest/regression_1000.c:1734: TEEC_InvokeCommand(&session, 16, ((void*)0), &ret_orig) has an unexpected value: 0xffff000c = TEEC_ERROR_OUT_OF_MEMORY, expected 0x0 = TEEC_SUCCESS
external/optee_test/host/xtest/regression_1000.c:1739: TEEC_InvokeCommand(&session, 17, ((void*)0), &ret_orig) has an unexpected value: 0xffff0000 = TEEC_ERROR_GENERIC, expected 0xffff3024 = TEE_ERROR_TARGET_DEAD
external/optee_test/host/xtest/regression_1000.c:1741: ret_orig has an unexpected value: 0x4 = TEEC_ORIGIN_TRUSTED_APP, expected 0x3 = TEEC_ORIGIN_TEE
  regression_1022 FAILED
```
```
E/LD:  init_elf:259 sys_open_ta_bin(b3091a65-9751-4784-abf7-0298a7cc35ba)
E/TC:? 0 call_ldelf_dlsym:691 ldelf dl_entry function panicked
E/TC:? 0 
E/TC:? 0 User TA data-abort at address 0x0 (translation fault)
E/TC:? 0  esr 0x92000005  ttbr0 0x800003f0cb080   ttbr1 0x00000000   cidr 0x0
E/TC:? 0  cpu #4          cpsr 0x00000000
E/TC:? 0  x0  0000000005880e93 x1  0000000000000000
E/TC:? 0  x2  0000000040005fd8 x3  0000000040010430
E/TC:? 0  x4  0000000000000000 x5  0000000040005fd8
E/TC:? 0  x6  0000000000000000 x7  000000003f0ccc90
E/TC:? 0  x8  000000003f0ccc8c x9  000000003f0ccc8c
E/TC:? 0  x10 000000008df7bae2 x11 0000000007736976
E/TC:? 0  x12 00000000de7e52b0 x13 0000000040005fc0
E/TC:? 0  x14 0000000040040270 x15 0000000071a791cb
E/TC:? 0  x16 000000003f022200 x17 000000004010a008
E/TC:? 0  x18 000000008e42f14e x19 0000000040005fc0
E/TC:? 0  x20 0000000000000000 x21 000000003f078ed0
E/TC:? 0  x22 0000000000000700 x23 0000000040005fe0
E/TC:? 0  x24 0000000040005fd8 x25 000000003f09b220
E/TC:? 0  x26 000000004010d000 x27 000000003f09b220
E/TC:? 0  x28 0000000000000000 x29 0000000040005f40
E/TC:? 0  x30 0000000040006150 elr 00000000400089c0
E/TC:? 0  sp_el0 0000000040005f40
E/LD:  Status of TA 5b9e0e40-2636-11e1-ad9e-0002a5d5c51b
E/LD:   arch: aarch64
E/LD:  region  0: va 0x40004000 pa 0x3f2f7000 size 0x002000 flags rw-s (ldelf)
E/LD:  region  1: va 0x40006000 pa 0x3f2f9000 size 0x007000 flags r-xs (ldelf)
E/LD:  region  2: va 0x4000d000 pa 0x3f300000 size 0x001000 flags rw-s (ldelf)
E/LD:  region  3: va 0x4000e000 pa 0x3f301000 size 0x003000 flags rw-s (ldelf)
E/LD:  region  4: va 0x40011000 pa 0x3f304000 size 0x001000 flags r--s
E/LD:  region  5: va 0x40012000 pa 0x00001000 size 0x00f000 flags r-xs [0]
E/LD:  region  6: va 0x40021000 pa 0x00010000 size 0x0e5000 flags rw-s [0]
E/LD:  region  7: va 0x40106000 pa 0x3f3f9000 size 0x003000 flags rw-s (stack)
E/LD:  region  8: va 0x40109000 pa 0x00000000 size 0x002000 flags r-xs [1]
E/LD:  region  9: va 0x4010b000 pa 0x00001000 size 0x001000 flags rw-s [1]
E/LD:  region 10: va 0x4010d000 pa 0x3f3ff000 size 0x001000 flags rw-s
E/LD:   [0] 5b9e0e40-2636-11e1-ad9e-0002a5d5c51b @ 0x40012000
E/LD:   [1] ffd2bded-ab7d-4988-95ee-e4962fff7154 @ 0x40109000
E/LD:   [2] b3091a65-9751-4784-abf7-0298a7cc35ba @ 0x00000000
E/LD:  Call stack:
E/LD:   0x00000000400089c0
```

Might have something to do with the recent ldelf patch. If (most probably) I can't find an obvious fix by tomorrow, please release 3.7.0 anyway and I'll resolve this by the next release. Thanks.

`Tested-by: Victor Chong <victor.chong@linaro.org> (Hikey620 AOSP except regression_1022)`

**Comment by jforissier** [comment]
@vchong NULL pointer in `ldelf`, you can probably find the source code location by looking in `ldelf.map`. I wonder why the panic code is `TEEC_ERROR_OUT_OF_MEMORY`.

**Comment by jbech-linaro** [comment]
Last chance to say whether you want something else to merged before I start tagging (which I will do in about an our from I post this).

@vchong , if you cannot resolve the AOSP issues, then we'll mention that in the release notes and tag a patch version when the issue has been resolved.

**Comment by vchong** [comment]
@jbech-linaro, ack, please go ahead. Will tag a patch version later.

**Comment by jforissier** [comment]
@vchong the first error it out of memory, have you tried increasing `CFG_CORE_HEAP_SIZE`? I often have to do it to pass some tests in some specific configurations as mentioned [here](https://docs.google.com/document/d/1EkgsyC-EZ2NnbfLAQkFOhPNFII3lFvwa8EIo3LFdGdk).

**Comment by vchong** [comment]
`Tested-by: Victor Chong <victor.chong@linaro.org> (Hikey620 AOSP)`

Sorry for the noise. There new TA that came with `os_test_lib_dl` that wasn't included in the build. Duh..

**Comment by clementfaure** [comment]
Hello,

Sorry for the late notice, but we have few issues with the following boards:
 * imx6sl where the CAAM driver should be disabled.
 * imx7ulp where the start base address is not correct.
 * imx6q series where there is a CAAM driver issue with ```CFG_REE_FS_TA_BUFFERED = n```

We have patches for these issues. The pull request will come this afternoon. Is it too late to include these patches in this RC?

Otherwise,
```
Tested-by: Clement Faure <clement.faure@nxp.com> (mx6ulevk, mx6ullevk)
Tested-by: Clement Faure <clement.faure@nxp.com> (mx6sllevk)
Tested-by: Clement Faure <clement.faure@nxp.com> (mx6sxsabresd)
Tested-by: Clement Faure <clement.faure@nxp.com> (mx7dsabresd)
Tested-by: Clement Faure <clement.faure@nxp.com> (mx8mm, mx8mn, mx8mq)

**Comment by jbech-linaro** [comment]
> Is it too late to include these patches in this RC?

I am in the middle of the process already, so yes, it's unfortunately too late for additional patches. But I can still squeeze in your Tested-By's.

**Pull request merged**